src/race.rs

```diff
@@ -50,6 +50,30 @@ impl OnceNonZeroUsize {
         NonZeroUsize::new(val)
     }
 
+    /// Get the reference to the underlying value, without checking if the cell
+    /// is initialized.
+    ///
+    /// # Safety
+    ///
+    /// Caller must ensure that the cell is in initialized state, and that
+    /// the contents are acquired by (synchronized to) this thread.
+    #[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
+    pub fn get_unchecked(&self) -> NonZeroUsize {
+        let p = self.inner.as_ptr();
+
+        // SAFETY: The caller is responsible for ensuring that the value
+        // was initialized and that the contents have been acquired by
+        // this thread. This method is only made available on a subset of
+        // targets where we are confident there will be no tearing or other
+        // interference with another thread that may be setting the value
+        // simultaneously.
+        let val = unsafe { p.read() };
```

Comment:
I think non-atomic read will be correct on any platform? The cell _must_ be initialized at this point, so, even if there are some concurrent `self.inner.compare_exchange`, they all will fail, and won't cause teared reads.

You can get the tear if you don't actually ensure the safety invariant here, and do `set/get_unchecked` literally in parallel, but that would be technical UB on x86 as well, from the perspective of Rust abstract machine.

Reply:
> even if there are some concurrent self.inner.compare_exchange, they all will fail, and won't cause teared reads.

I will think about this one overnight. I have MSRV 1.63.0 in my projects so I might try to find an alternative.

Reply:
I wonder if just doing Relaxed load could optimize well?

Reply:
I tested it in my codebase. The relaxed load generates the same exact code as using `get()`. The code as presented here avoids subsequent reloads after the first call to `get()`.